### DISCUSSION TITLE
Enumeration trees using Merkle Regs

### DISCUSSION THREAD

**Discussion by DaveLanday** [spacejam/rust-crdt#149] [Q&A] [answered]
I am learning about Merkle-DAG CRDTs and think they might be applicable to my use case. Let's say I have a computation that is distributed and the computation in question requires some load balancing since work during the computation can be passed off. Imagine that the computation is enumerable in the sense that at each iteration, I use the information from prior computations to solve the current one. Let's also assume that from the information of prior computations, I can create two potential branches of work to do (i.e: my prior work says that I should try the case where the variable x can be $\geq$ 6, or $\leq$ 5 after determining that variable y=10) and I need to solve both in order to know which will produce the next set of work to do (i.e: if I found that x $\geq$ 6 yields a better result and now I need to solve for z $\geq$ 4 or z $\leq$ 5). If I want to solve the case where x $\geq$ 6 locally and send the other bit of work (x $\leq$ 5) off to a remote worker, then the remote worker should be able to access the information from the prior computations (y = 10). Would a Merkle-DAG be appropriate to represent the history of the computation? Would a Merkle-Reg be appropriate to store the values of the enumerable information at each step in the computation (i.e: if I found that y=10 at iter 1, x=6 at iter 2, j=3 at iter 3, etc...)? Any direction here would help. Thanks!

**Comment by davidrusu** [comment, marked as answer]
Yes, you could use the merkle reg to store the intermediate results of the computation.

The concurrent values in the Merkle Register could represent the next step of the computation.

Each node when it wants to do some work, could read off the current values in the register, pick off one of the values, run the next step based on the value representing the result of the prior computation.

Then replace that value in the merkle register with the results to be explored in next iteration.

```rust
struct Task(String)

let mut merkle_reg = MerkleReg::new()
let initial_task = Task("Find x, y, z s.t. f(x, y, z) = 0")
let initial_node = merkle_reg.write(initial_task.clone(), BTreeSet::new());

// --- the node is distributed to the workers


// ----- worker with it's own copy of the merkle reg

if let Some(task_node) = merkle_reg.read().nodes.first() {
    // does the task
    let follow_up_tasks = do_task(task_node.val);

    // then it writes the followup tasks with references to the task that spawned these followups.
    let follow_up_nodes = follow_up_tasks
        .into_iter()
        .map(|t| merkle_reg.write(t, BtreeSet::from_iter([task_node.hash()]));
    // worker than distributes these follow up nodes to other workers
}
```


Some work needs to be done to decide which of the concurrent tasks to do when you read from the merkle register, but something like that should work.

**Reply by DaveLanday** [reply]
Wow! Thank you for the reply. Potentially, I could store the worker a task is assigned to in the merkle register as well? For example assign a worker a unique id at runtime, perhaps use a main process thread to create the assignments for each task. Workers get the update and read from the node(s) with their assignment. The other way is more traditional where the main thread has a queue and workers pull items off of it independently and then can sign them before updating the dag after completing the task.

**Reply by davidrusu** [reply]
If you want to take full advantage of CRDT's you need to design a work assignment mechanism that uses only local information (ie. no main process assigning work).

If tasks are deterministic, and you don't mind a bit of redundancy, you could have workers select a random task from the register, some workers may redo some computation of another task but it has other benefits in that it makes the system resilient to any worker crashing. (contrast to the case where you assign workers to tasks up front, if that worker dies, then that sub-tree of the computation will stall.)

Just some ideas. You can experiment with different approaches to see what works best for you.

**Reply by DaveLanday** [reply]
Thanks for this awesome feedback! I really appreciate you taking the time @davidrusu
